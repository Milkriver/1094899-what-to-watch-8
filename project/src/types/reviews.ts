export interface IReview {
    id: number,
    movieId: number,
    author: string,
    date: string,
    rating: number,
    text: string,
  }
