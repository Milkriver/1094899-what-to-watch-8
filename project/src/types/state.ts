import { AuthorizationStatus } from '../const';
import { IMovie } from './common';
import { IReviewResponse } from './reviews';

export interface IState {
  currentGenre: string
  movies: IMovie[]
  isDataLoaded: boolean
  authorizationStatus: AuthorizationStatus,
  movie: IMovie,
  sameMovies: IMovie[],
  reviews: IReviewResponse[],
  promoMovie: IMovie,
}
