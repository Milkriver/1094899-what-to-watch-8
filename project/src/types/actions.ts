import { AxiosInstance } from 'axios';
import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { IState } from './state';

export enum ActionType {
  ChangeGenre = 'movieList/ChangeGenre',
  LoadMovies = 'movieList/LoadMovies',
  LoadSingleMovie = 'movieCard/LoadSingleMovie',
  LoadActiveMovie = 'movieCard/LoadActiveMovie',
  LoadSameGenreMovies = 'movieCard/LoadSameGenreMovies',
  LoadPromoMovie = 'movieCard/LoadPromoMovie',
  LoadFavouriteMovies = 'userList/LoadFavouriteMovies',
  AddFavoriteMovie = 'userList/AddFavoriteMovie',
  RemoveFavoriteMovie = 'userList/RemoveFavoriteMovie',
  LoadReviews = 'movieCard/LoadReviews',
  AddReview = 'movieCard/AddReview',
  RequireAuthorization = 'user/requireAuthorization',
  RequireLogout = 'user/requireLogout',
  LoadUserData = 'user/loadUserData',
}

export type ThunkActionResult<R = Promise<void>> = ThunkAction<R, IState, AxiosInstance, Action>;

export type ThunkAppDispatch = ThunkDispatch<IState, AxiosInstance, Action>;
