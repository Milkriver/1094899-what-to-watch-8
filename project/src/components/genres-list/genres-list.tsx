import { connect, ConnectedProps } from 'react-redux';
import { Dispatch } from 'redux';
import { changeGenre } from '../../store/action';
import { Actions } from '../../types/actions';
import { IMovie } from '../../types/common';
import { IState } from '../../types/state';

interface IGenresListProps {
  movies: IMovie[],
}

const mapStateToProps = ({ currentGenre }: IState) => ({
  currentGenre,
});

const mapDispatchToProps = (dispatch: Dispatch<Actions>) => ({
  onChangeGenre(genre: string) {
    dispatch(changeGenre(genre));
  },
});

const connector = connect(mapStateToProps, mapDispatchToProps);
type PropsFromRedux = ConnectedProps<typeof connector>;
type ConnectedComponentProps = PropsFromRedux & IGenresListProps;


function GenresList(props: ConnectedComponentProps): JSX.Element {
  const { movies, currentGenre, onChangeGenre } = props;
  const genreList = Array.from(new Set(movies.map((movieElement) => movieElement.genre)));
  const genreListMenu = ['All genres', ...genreList];

  return (
    <ul className="catalog__genres-list">
      {genreListMenu.map((genre: string) => (
        <li className={currentGenre === genre ? 'catalog__genres-item catalog__genres-item--active' : 'catalog__genres-item'}
          key={genre}
        >
          <div className="catalog__genres-link " onClick={() => onChangeGenre(genre)}>
            {genre}
          </div>
        </li>
      ))}
    </ul>

  );
}
export { GenresList };
export default connector(GenresList);
