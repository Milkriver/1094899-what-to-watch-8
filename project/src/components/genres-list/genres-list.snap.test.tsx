import { render } from '@testing-library/react';
import { GenresList } from './genres-list';
import { IMovie } from '../../types/common';

const mockMovie1: IMovie = {
  id: 1, name: 'name1', posterImage: 'posterImage1', previewImage: 'previewImage1', backgroundImage: 'backgroundImage1',
  backgroundColor: 'black', videoLink: 'videoLink1', previewVideoLink: 'previewVideoLink1',
  description: 'description1', rating: 1, scoresCount: 1, director: 'director1',
  starring: ['actor1', 'actor2'], runTime: 1, genre: 'Comedy', released: 1, isFavorite: true,
};

const mockMovie2: IMovie = {
  id: 2, name: 'name2', posterImage: 'posterImage2', previewImage: 'previewImage2', backgroundImage: 'backgroundImage2',
  backgroundColor: 'black', videoLink: 'videoLink2', previewVideoLink: 'previewVideoLink2',
  description: 'description2', rating: 2, scoresCount: 2, director: 'director2',
  starring: ['actor1', 'actor2'], runTime: 2, genre: 'Comedy', released: 2, isFavorite: true,
};

const mockMovies: IMovie[] = [mockMovie2, mockMovie1];

describe('Component: GenresList', () => {
  test('should render correctly', () => {
    const mockGenre = 'Comedy';
    const mockFn = () => Promise.resolve();
    const page = <GenresList currentGenre={mockGenre} movies={mockMovies} onChangeGenre={mockFn} />;
    const { container } = render(page);
    expect(container).toMatchSnapshot();
  });
});
