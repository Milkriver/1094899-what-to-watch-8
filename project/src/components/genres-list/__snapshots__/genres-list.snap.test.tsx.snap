// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: GenresList should render correctly 1`] = `
<div>
  <ul
    class="catalog__genres-list"
  >
    <li
      class="catalog__genres-item"
    >
      <div
        class="catalog__genres-link "
      >
        All genres
      </div>
    </li>
    <li
      class="catalog__genres-item catalog__genres-item--active"
    >
      <div
        class="catalog__genres-link "
      >
        Comedy
      </div>
    </li>
  </ul>
</div>
`;
